### PR TITLE
Add notebooks with example usage of arithmetic ops

### PR DESCRIPTION
**Pull request by klecki**
Adjust Cast to support all types that we support.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
It adds better jupyter notebooks showing the example usage of arithmetic ops.

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
N/A
 - What was changed, added, removed?
The notebooks were addded, cast was fixed.
 - What is most important part that reviewers should focus on?
Cast adjustments, the narrative in the notebooks.
 - Was this PR tested? How?
Notebooks tested locally, CI
 - Were docs and examples updated, if necessary?
This is basically about adding documentation.

**JIRA TASK**: [DALI-1132]

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

File: docs/examples/expressions/expr_type_promotions.ipynb [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+   "metadata": {},
+   "source": [
+    "### Defining the pipeline\n",
+    "\n",
+    "The next step is to define the Pipeline. We override `Pipeline.iter_setup`, a method called by the pipeline before every `Pipeline.run`. It is meant to feed the data into `ExternalSource()` operators indicated by `self.left` and `self.right`.\n",
+    "The data will be obtained from `get_data` function to which we pass the left and right types. \n",
+    "\n",
+    "Note, that we do not need to instantiate any additional operators, we can use regular Python arithmetic expression on the results of other operators in the `define_graph` step.\n",
+    "\n",
+    "For convenience, we'll wrap the usage of arithmetic operations in a lambda called `operation`, specified when creating the pipeline.\n",
```
Comment (JanuszL) [inline docs/examples/expressions/expr_type_promotions.ipynb]: While lambda provides an easy way to parameterize example it only obfuscates the fact that you can just call arithmetic op on the graph edges (it is shown in expr_blend_image.ipynb but one needs to start reading examples from it).
Maybe in the top paragraph, you can point to a more basic example in `expr_blend_image.ipynb` and say that this one shows more edge cases and is rather an advanced tutorial?

File: include/dali/core/convert.h [hunk 2]
```diff
@@ -77,14 +77,37 @@ struct ret_type {  // a placeholder for return type
...
-    needs_clamp<U, T>::value && std::is_signed<U>::value,
+    needs_clamp<U, T>::value && std::is_signed<U>::value && std::is_signed<T>::value,
     T>
 clamp(U value, ret_type<T>) {
   return value <= min_value<T>() ? min_value<T>() :
          value >= max_value<T>() ? max_value<T>() :
          static_cast<T>(value);
 }
 
+template <typename T, typename U>
```
Comment (jantonguirao) [inline include/dali/core/convert.h]: This one is exacly the same as the one above. Was it needed then?

File: docs/examples/expressions/cats.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+kitten/cat_1.jpg 1
```
Comment (jantonguirao) [inline docs/examples/expressions/cats.txt]: Curious, is it possible to just let the reader get all the files in the subfolders?

File: qa/TL0_jupyter/test.sh [hunk 4]
```diff
@@ -13,7 +13,7 @@ test_body() {
     # dummy patern
     black_list_files="multigpu"
 
-    ls *.ipynb | sed "/${black_list_files}/d" | xargs -i jupyter nbconvert \
+    ls *.ipynb expressions/*.ipynb | sed "/${black_list_files}/d" | xargs -i jupyter nbconvert \
```
[the remark this answers is not in the document]
Comment (klecki) [reply qa/TL0_jupyter/test.sh]: I didn't know about the TL1 test.

File: docs/examples/expressions/cats.txt (hunk 3 above)
Comment (klecki) [reply docs/examples/expressions/cats.txt]: It assumes some structure of `root_dir/class_dir`. And I want two explicit batches that are separate (cats vs dogs).

File: dali/operators/util/cast.h [hunk 5]
```diff
@@ -22,6 +22,10 @@
 
 namespace dali {
 
+
+#define CAST_ALLOWED_TYPES \
+  (uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float16, float, double)
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/operators/util/cast.h]: Good spot.

**Comment by klecki** [comment]
> In `expr_blend_image.ipynb`:
> 
> > Here we used python immidiate values as constant inputs to our arithmetic expression.
> 
> immediate
> 
> > Lets create an instance of the pipelie and build it. We use batch_size = 1 for simplicity of showing the result.
> 
> Let's, Pipeline
> 
> > Now, we wil run and display the results. You can play this cell several times to see the result for different images.
> 
> will
> 
> Also for `(0.4 * cats + 0.6 * dogs) / 2`, why dividing by 2 to when blending? `0.4 * cats + 0.6 * dogs` should be enough

Done.

File: include/dali/core/convert.h (hunk 2 above)
Comment (klecki) [reply include/dali/core/convert.h]: There was an issue for some T unsigned case that need special handling, as compiler was complaining for signed/unsigned comparisons, you also cannot make and fp type unsigned.

File: docs/examples/expressions/expr_type_promotions.ipynb (hunk 1 above)
Comment (klecki) [reply docs/examples/expressions/expr_type_promotions.ipynb]: I split it into two tutorials, the problem is the data setup is a bit repetitive now.

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Comment by ptrendx** [comment]
> > Also for `(0.4 * cats + 0.6 * dogs) / 2`, why dividing by 2 to when blending? `0.4 * cats + 0.6 * dogs` should be enough
> 
> Done.

Hmm, does matplotlib do some sort of normalization? Because the float and uint16 versions of the image look the same and the float one is normalized while uint16 still has the half factor in it, so should be darker/less saturated...

**Comment by ptrendx** [comment]
Actually nevermind, I got confused by the `Done` note - it seems that is not done yet :-).

**Comment by JanuszL** [comment]
Still why you divise by 2?
```
blend_float = (0.4 * cats + 0.6 * dogs) / 2
```
Typo:
> signle constant

**Comment by klecki** [comment]
> Still why you divise by 2?

I just started that way, it's not necessary but shows a more complex formula. I can remove it if you want. I adjusted the description to fit the division by 2. Should I remove it?

**Comment by banasraf** [comment]
Some nitpicks 
In expr_examples.ipynb

> They can be used as operation between two tensors (returned by some operators)

This parenthesis sounds a bit like it works for outputs of _only some_ operators. I would say this information in parenthesis isn't needed at all but I understand you wanted to say that it works for `EdgeReferences` and not actual tensors.

**Comment by JanuszL** [comment]
> > Still why you divise by 2?
> 
> I just started that way, it's not necessary but shows a more complex formula. I can remove it if you want. I adjusted the description to fit the division by 2. Should I remove it?

Leave it as it is.

File: docs/examples/expressions/expr_examples.ipynb [hunk 6]
```diff
@@ -0,0 +1,508 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# DALI expressions and arithmetic operators\n",
+    "\n",
+    "In this example, we will show simple examples how to use arithmetic operators in DALI Pipeline that allow for elementwise operations on tensors inside a pipeline. We will show available operators and examples of using constant and scalar inputs.\n",
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    "\n",
+    "## Supported operators\n",
+    "\n",
+    "DALI currently supports binary arithmetic operators, namely: `+`, `-`, `*`, `/` and `//`. They can be used as operation between two tensors (returned by some operators), between tensor and a scalar or tensor and a constant."
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Prepare the test pipeline\n",
+    "\n",
+    "First, we will prepare helper code, so we can easily manipulate the types and values that will appear as tensors in DALI pipeline.\n",
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+    "\n",
+    "We use `from __future__ import division` to allow `/` and `//` as true division and floor division operators.\n",
+    "We will be using numpy as source for the custom provided data and we also need to import several things from DALI needed to create Pipeline and use ExternalSource Operator. "
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+    "examine_output(out)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As we can see the resulting tensors are obtained by applying the arithmetic operation between corresponding elements of its inputs.\n",
+    "\n",
+    "The shapes of the arguments to arithmetic operators should match (with an exception for scalar tensor inputs that we will describe in the next section) otherwise we will get an error."
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Constant and scalar operands\n",
+    "Untill now we considered only tensor inputs of matching shapes for inputs of arithmetic operators. DALI allows one of the operands to be a constant or a tensor consisting of scalars. They can appear on both sides of binary expression.\n",
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

File: docs/examples/expressions/expr_examples.ipynb [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+    "As we can see the constant operand is broadcasted to all elements of all tensors in the batch."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Scalars\n",
+    "\n",
+    "As an addition to elementwise operation between Tensors of the same sizes, DALI allows to use tensors containing scalar values - that is a batch of elements with shape `{1}` as one of the operands. The tensor of scalars will behave similarly to a constant described above - each scalar value will be broadcasted to every element of the other tensor operand. Note that contrary to constants, the tensor is still as a batch of elements - each scalar value is broadcasted to corresponding tensor.\n",
```
[the remark this answers is not in the document]
Comment (klecki) [reply docs/examples/expressions/expr_examples.ipynb]: Done

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
> nitpick
> In expr_examples.ipynb
> 
> > They can be used as operation between two tensors (returned by some operators)
> 
> This parenthesis sounds a bit like it works for outputs of _only some_ operators. I would say this information in parenthesis isn't needed at all but I understand you wanted to say that it works for `EdgeReferences` and not actual tensors.

Done

**Pull request merged**